src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,369 @@
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	uint8_t *config;
+	size_t size;
+	int ret;
+
+	comp_dbg(dev, "google_ctc_audio_processing_reconfigure()");
+
+	if (!comp_is_current_data_blob_valid(cd->tuning_handler) &&
+	    !comp_is_new_data_blob_available(cd->tuning_handler)) {
+		/*
+		 * The data blob hasn't been available once so far.
+		 *
+		 * This looks redundant since the same check will be done in
+		 * comp_get_data_blob() below. But without this early return,
+		 * hundreds of warn message lines are produced per second by
+		 * comp_get_data_blob() calls until the data blob is arrived.
+		 */
+		return 0;
+	}
+
+	config = comp_get_data_blob(cd->tuning_handler, &size, NULL);
+	if (size == 0) {
+		/* No data to be handled */
+		return 0;
+	}
+
+	if (!config) {
+		comp_err(dev, "google_ctc_audio_processing_reconfigure(): Tuning config not set");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "google_ctc_audio_processing_reconfigure(): New tuning config %p (%zu bytes)",
+		  config, size);
+
+	cd->reconfigure = false;
+
+	uint8_t *google_ctc_audio_processing_config;
+	size_t google_ctc_audio_processing_config_size;
+	int num_channels;
+	bool google_ctc_audio_processing_config_present;
+	bool num_channels_present;
+
+	GoogleCtcAudioProcessingParseSofConfigMessage(config, size,
+						      &google_ctc_audio_processing_config,
+						      &google_ctc_audio_processing_config_size,
+						      &num_channels,
+						      &google_ctc_audio_processing_config_present,
+						      &num_channels_present);
+
+	if (google_ctc_audio_processing_config_present) {
+		comp_info(dev,
+			  "google_ctc_audio_processing_reconfigure(): Applying config of size %zu bytes",
+			  google_ctc_audio_processing_config_size);
+
+		ret = GoogleCtcAudioProcessingReconfigure(cd->state,
+							  google_ctc_audio_processing_config,
+							  google_ctc_audio_processing_config_size);
+		if (ret) {
+			comp_err(dev, "GoogleCtcAudioProcessingReconfigure failed: %d",
+				 ret);
+			return ret;
+		}
+	}
+
+	if (num_channels_present) {
+		cd->num_channels = num_channels;
+		comp_info(dev,
+			  "google_ctc_audio_processing_reconfigure(): Applying num channels %d",
+			  cd->num_channels);
+	}
+
+	return 0;
+}
+
+static int google_ctc_audio_processing_prepare(struct processing_module *mod)
+{
+	struct comp_dev *dev = mod->dev;
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	comp_info(dev, "google_ctc_audio_processing_prepare()");
+
+	return 0;
+}
+
+static int google_ctc_audio_processing_reset(struct processing_module *mod)
+{
+	comp_dbg(mod->dev, "google_ctc_audio_processing_reset()");
+
+	return 0;
+}
+
+static int google_ctc_audio_processing_process(struct processing_module *mod,
+					       struct input_stream_buffer *input_buffers,
+					       int num_input_buffers,
+					       struct output_stream_buffer *output_buffers,
+					       int num_output_buffers)
```

Comment:
This is implementing the older process_audio_stream API, not the newer process() one which works with the sink/source API.  I got dinged on this with the AEC PR too, in theory either can still be used, but sink/source is futureproof.

Reply:
Changed to the source/sink API.